From 57793d9e81c8cde68d4083117602effa69b5045e Mon Sep 17 00:00:00 2001
Message-Id: <57793d9e81c8cde68d4083117602effa69b5045e.1424394676.git.feitong.yi@intel.com>
In-Reply-To: <00f4430a265007e8e2c016cc7926bf530835ea30.1424394676.git.feitong.yi@intel.com>
References: <00f4430a265007e8e2c016cc7926bf530835ea30.1424394676.git.feitong.yi@intel.com>
From: Akashdeep Sharma <akashdeep.sharma@intel.com>
Date: Mon, 2 Feb 2015 10:15:04 +0530
Subject: [PATCH 06/61] MUST_REBASE [VPG]: video/adf: Extcon class
 notification

Extcon class driver HDMI notification to inform userspace about
HDMI hot plug/ unplug.

MUST_REBASE: Since Google ADF framework is not in upstream and also
we cannot have two display drivers, other one being i915, we need to
work on i915 and ADF convergence path before it can be upstreamed.

Issue: GMINL-5532
Change-Id: Ic831860a7da6c17fd7906fbe235cd2effe392b97
Signed-off-by: Akashdeep Sharma <akashdeep.sharma@intel.com>
---
 .../adf/intel/core/common/hdmi/gen_hdmi_hotplug.c  |    8 ++++++++
 .../adf/intel/core/common/hdmi/gen_hdmi_pipe.c     |    9 +++++++++
 .../intel/include/core/common/hdmi/gen_hdmi_pipe.h |    5 +++++
 3 files changed, 22 insertions(+)

diff --git a/drivers/video/adf/intel/core/common/hdmi/gen_hdmi_hotplug.c b/drivers/video/adf/intel/core/common/hdmi/gen_hdmi_hotplug.c
index 3a6768f..7fa9a10 100644
--- a/drivers/video/adf/intel/core/common/hdmi/gen_hdmi_hotplug.c
+++ b/drivers/video/adf/intel/core/common/hdmi/gen_hdmi_hotplug.c
@@ -629,8 +629,16 @@ bool hdmi_notify_audio(struct hdmi_pipe *hdmi_pipe, bool connected)
 	if (connected) {
 		if (hdmi_pipe->config.ctx.monitor->has_audio)
 			adf_hdmi_audio_signal_event(HAD_EVENT_HOT_PLUG);
+#ifdef CONFIG_EXTCON
+		if (strlen(hdmi_pipe->hotplug_switch.name) != 0)
+			extcon_set_state(&hdmi_pipe->hotplug_switch, 1);
+#endif
 	} else {
 		adf_hdmi_audio_signal_event(HAD_EVENT_HOT_UNPLUG);
+#ifdef CONFIG_EXTCON
+		if (strlen(hdmi_pipe->hotplug_switch.name) != 0)
+			extcon_set_state(&hdmi_pipe->hotplug_switch, 0);
+#endif
 	}
 
 	return true;
diff --git a/drivers/video/adf/intel/core/common/hdmi/gen_hdmi_pipe.c b/drivers/video/adf/intel/core/common/hdmi/gen_hdmi_pipe.c
index adac57e..b05bfe3 100644
--- a/drivers/video/adf/intel/core/common/hdmi/gen_hdmi_pipe.c
+++ b/drivers/video/adf/intel/core/common/hdmi/gen_hdmi_pipe.c
@@ -170,6 +170,10 @@ void hdmi_config_destroy(struct hdmi_config *config)
 
 void hdmi_pipe_destroy(struct hdmi_pipe *pipe)
 {
+#ifdef CONFIG_EXTCON
+	if (pipe)
+		extcon_dev_unregister(&pipe->hotplug_switch);
+#endif
 	if (pipe)
 		hdmi_config_destroy(&pipe->config);
 	pr_info("ADF: HDMI: %s", __func__);
@@ -211,6 +215,11 @@ int hdmi_pipe_init(struct hdmi_pipe *pipe,
 
 	adf_hdmi_audio_init(pipe);
 
+#ifdef CONFIG_EXTCON
+	pipe->hotplug_switch.name = "hdmi";
+	extcon_dev_register(&pipe->hotplug_switch);
+#endif
+
 	/*
 	 * Configure HDMI
 	 * This function can probe HDMI and try to do a modeset
diff --git a/drivers/video/adf/intel/include/core/common/hdmi/gen_hdmi_pipe.h b/drivers/video/adf/intel/include/core/common/hdmi/gen_hdmi_pipe.h
index 0545ce1..f908619 100644
--- a/drivers/video/adf/intel/include/core/common/hdmi/gen_hdmi_pipe.h
+++ b/drivers/video/adf/intel/include/core/common/hdmi/gen_hdmi_pipe.h
@@ -19,6 +19,8 @@
 
 #include "core/intel_dc_config.h"
 #include "core/common/hdmi/gen_hdmi_audio.h"
+#include <linux/extcon.h>
+#include <linux/types.h>
 
 #define HDMI_MAX_ELD_LENGTH	84
 #define HDMI_DIP_PACKET_HEADER_LEN	3
@@ -141,6 +143,9 @@ struct hdmi_pipe {
 
 	/* Added for HDMI audio */
 	uint32_t tmds_clock;
+#ifdef CONFIG_EXTCON
+	struct extcon_dev hotplug_switch;
+#endif
 };
 
 static inline struct hdmi_pipe *
-- 
1.7.9.5

